### PR TITLE
Refactor CUDA block size deduction

### PR DESCRIPTION
**Pull request by dalg24**
This is an attempt to cleanup block size deduction and get rid of duplicated code.

### CODE REVIEW

File: core/src/Cuda/Kokkos_Cuda_BlockSize_Deduction.hpp [hunk 1]
```diff
@@ -48,18 +48,134 @@
...
+      case 6: return 32;
+      case 7: {
+        int isTuring = properties.minor == 5;
+        return (isTuring) ? 16 : 32;
+      }
+      default:
+        throw_runtime_exception("Unknown device in cuda block size deduction");
+    }
+  }();
+#endif
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_BlockSize_Deduction.hpp:119]: @crtrott My intent is to account for this in `cuda_max_active_blocks_per_sm()` but I did not want to change too much of the current logic at once.

**Comment by dalg24** [comment]
FYI `develop` has
```diff
diff -u cuda_get_max_block_size cuda_get_opt_block_size
--- cuda_get_max_block_size	2020-06-17 10:35:09.000000000 -0400
+++ cuda_get_opt_block_size	2020-06-17 10:35:33.000000000 -0400
@@ -1,5 +1,5 @@
 template <class FunctorType, class LaunchBounds>
-int cuda_get_max_block_size(const CudaInternal* cuda_instance,
+int cuda_get_opt_block_size(const CudaInternal* cuda_instance,
                             const cudaFuncAttributes& attr,
                             const FunctorType& f, const size_t vector_length,
                             const size_t shmem_block,
@@ -37,12 +37,9 @@
   }
   int opt_block_size = (blocks_per_sm >= min_blocks_per_sm) ? block_size : 0;
   int opt_threads_per_sm = threads_per_sm;
-  // printf("BlockSizeMax: %i Shmem: %i %i %i %i Regs: %i %i Blocks: %i %i
-  // Achieved: %i %i Opt: %i %i\n",block_size,
-  //   shmem_per_sm,max_shmem_per_block,functor_shmem,total_shmem,
-  //   regs_per_sm,regs_per_thread,max_blocks_shmem,max_blocks_regs,blocks_per_sm,threads_per_sm,opt_block_size,opt_threads_per_sm);
+
   block_size -= 32;
-  while ((blocks_per_sm == 0) && (block_size >= 32)) {
+  while ((block_size >= 32)) {
     functor_shmem =
         FunctorTeamShmemSize<FunctorType>::value(f, block_size / vector_length);
     total_shmem = shmem_block + shmem_thread * (block_size / vector_length) +
@@ -65,10 +62,6 @@
         opt_threads_per_sm = threads_per_sm;
       }
     }
-    // printf("BlockSizeMax: %i Shmem: %i %i %i %i Regs: %i %i Blocks: %i %i
-    // Achieved: %i %i Opt: %i %i\n",block_size,
-    //   shmem_per_sm,max_shmem_per_block,functor_shmem,total_shmem,
-    //   regs_per_sm,regs_per_thread,max_blocks_shmem,max_blocks_regs,blocks_per_sm,threads_per_sm,opt_block_size,opt_threads_per_sm);
     block_size -= 32;
   }
   return opt_block_size;
```

**Comment by crtrott** [comment]
Super nervous about touching this. But After going through the code a couple times I do believe it should give the same answer as before.

**Comment by dalg24** [comment]
@crtrott should I go ahead and remove the old code behind the macro?

**Comment by crtrott** [comment]
yes

**Pull request merged**